google-cloud-spanner/src/main/java/com/google/cloud/spanner/AsyncResultSetImpl.java

```diff
@@ -593,4 +627,28 @@ public Struct getCurrentRowAsStruct() {
     checkValidState();
     return currentRow;
   }
+
+  @Override
+  public void onStreamMessage(
+      PartialResultSet partialResultSet,
+      int prefetchChunks,
+      int currentBufferSize,
+      StreamMessageRequestor streamMessageRequestor) {
+    synchronized (monitor) {
+      if (produceRowsInitiated) {
+        return;
+      }
+      // if PartialResultSet contains resume token or buffer size is more than configured size or
+      // we have reached end of stream, we can start the thread
+      boolean startJobThread =
+          !partialResultSet.getResumeToken().isEmpty()
+              || currentBufferSize >= prefetchChunks
+              || partialResultSet == GrpcStreamIterator.END_OF_STREAM;
+      if (startJobThread || state != State.IN_PROGRESS) {
+        initiateProduceRows();
+      } else {
+        streamMessageRequestor.requestMessages(1);
```

Comment:
Should we really do this? If I understand the code correctly, then the following will happen in most cases:
1. The stream is started with prefetchChunks=4. This means that we have already requested 4 messages.
2. The first time that this callback is triggered, the chunk is likely to have an empty resume token. In addition, the buffer is not full, and it is not the END_OF_STREAM. So this means that we request one more message (although we already had requested 4, and the buffer size is 4)
3. Step 2 repeats for chunk 2 and 3. Meaning that we have now requested in total 3 more chunks in addition to the first 4.
4. The fourth chunk contains a resume token and the ProduceRowsRunnable starts. This again will start removing data from the buffer, and each time a chunk is removed from the gRPC buffer, another message will be requested. Meaning that we continuously have requested 3 chunks more than that the buffer can contain.

Or am I misunderstanding how this will behave?

Reply:
You are right. This is just an improvement to have some extra chunks in the buffer.